test/test-sortedarray.c

```diff
@@ -0,0 +1,241 @@
+/*
+
+Copyright (c) 2016, Stefan Cloudt
+
+Permission to use, copy, modify, and/or distribute this software
+for any purpose with or without fee is hereby granted, provided
+that the above copyright notice and this permission notice appear
+in all copies.
+
+THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL
+WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE
+AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR
+CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
+LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT,
+NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
+CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <assert.h>
+#include <time.h>
+
+#include "alloc-testing.h"
+#include "framework.h"
+
+#include "sortedarray.h"
+
+#define TEST_SIZE 100
+
+int compare_float(void* v1, void* v2) 
```

Comment:
I understand what you mean. The sortedarray structure supports separate comparison and equality functions. So I use floats to make the non fractional part used for sorting and the fractional part for equality. My index of is structured in such way that it needs to find the equal object, not the object which compares to be equal. Since I do not do any arithmetic with floats comparisons won't be a problem, however it is indeed wiser to use ints since it is less likely to introduce a bug. However to keep the special case for elements which compare equal which are not equal I would not be able to use the already built in functions.